### PR TITLE
Arithmetic float operations

### LINKED ISSUE #241: Add support for floating point arithmetics and operations
**Description:**
Add support for Float arithmetic (addition, subtraction, multiplication, division, unary ops, etc) and Float comparison ( greater or equal, etc).

**Describe your problem(s)**
For a simple `a + b`, `add` instruction is generated instead of `fadd`.

for example - `add float %"%1_temp", %"%3_temp"`

**Describe your solution(s)**
Generate instructions such as `fadd float` in the CodeGen phase.

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by ramakotareddy**
Support for Float Arithmetic Operations

Resolves #241 

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

**Comment by ramakotareddy** [comment]
Fixed all the review comments from #250. Created separate issue for sign inversion and other missing float arithmetic operation test cases #270.

File: compiler/codegen/BinaryOpInsn.cpp [hunk 1]
```diff
@@ -36,67 +36,134 @@ void BinaryOpInsn::translate(llvm::Module &module, llvm::IRBuilder<> &builder) {
     auto *lhsRef = funcObj.getLLVMLocalOrGlobalVar(getLhsOperand(), module);
     auto *rhsOp1ref = funcObj.createTempVariable(rhsOp1, module, builder);
     auto *rhsOp2ref = funcObj.createTempVariable(rhsOp2, module, builder);
+    TypeTag rhsType = funcObj.getLocalOrGlobalVariable(rhsOp1).getType().getTypeTag();
+    auto *binaryOpResult = (llvm::Value *)nullptr;
```
Comment (ruvi-d) [inline compiler/codegen/BinaryOpInsn.cpp:40]: Suggest to place with: `llvm::Value *binaryOpResult = nullptr;`

File: compiler/codegen/BinaryOpInsn.cpp [hunk 2]
```diff
@@ -36,67 +36,134 @@ void BinaryOpInsn::translate(llvm::Module &module, llvm::IRBuilder<> &builder) {
...
         break;
     }
     case INSTRUCTION_KIND_BINARY_GREATER_THAN: {
-        auto *ifReturn = builder.CreateICmp(llvm::CmpInst::Predicate::ICMP_SGT, rhsOp1ref, rhsOp2ref, lhstmpName);
-        funcObj.insertBranchComparisonValue(lhsName, ifReturn);
-
+        if (rhsType == nballerina::TYPE_TAG_INT) {
+            binaryOpResult = builder.CreateICmp(llvm::CmpInst::Predicate::ICMP_SGT, rhsOp1ref, rhsOp2ref, lhstmpName);
+        } else if (rhsType == nballerina::TYPE_TAG_FLOAT) {
+            binaryOpResult = builder.CreateFCmp(llvm::CmpInst::Predicate::FCMP_OGT, rhsOp1ref, rhsOp2ref, lhstmpName);
```
Comment (ruvi-d) [inline compiler/codegen/BinaryOpInsn.cpp:102]: Any reason to pic `FCMP_OGT` over `FCMP_UGT`? Better have tests to cover NaN if we are doing OGT and if that is inline with Ballerina spec. 
Ditto for other uses of floating point ordered instruction usage.

File: test/float/add_float_negative_values.bal [hunk 3]
```diff
@@ -0,0 +1,15 @@
+// RUN: "%testRunScript" %s %nballerinacc "%java_path" "%target_variant" "%skip_bir_gen" | filecheck %s
```
Comment (ruvi-d) [inline test/float/add_float_negative_values.bal:1]: I think it's ok to move this inside `add_float.bal`
Ditto for the other separate negative value test cases.

File: test/float/greater_equal_than_float.bal [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+public function main() {
+    float a = 10.5;
+    float b = 5.5;
+    print_string("RESULT=");
+    if (a >= b) {
+        printf64(a);
+    }
+    else {
+        printf64(b);
+    }
```
Comment (ruvi-d) [inline test/float/greater_equal_than_float.bal:16]: Need to test the equal case too

**Review by ruvi-d — CHANGES_REQUESTED** [review]

File: compiler/codegen/BinaryOpInsn.cpp (hunk 2 above)
Comment (ramakotareddy) [reply compiler/codegen/BinaryOpInsn.cpp:102]: @ruvi-d, there is no particular reason to use  FCMP_OGT, can i use FCMP_UGT instead if FCMP_OGT ?

Comment (ruvi-d) [reply compiler/codegen/BinaryOpInsn.cpp:102]: @ramakotareddy I am expecting you to figure out what the valid float values are by referring to the Ballerina spec and/or jBallerina implementation. Accordingly, you can pick the ordered or unordered float instructions. If you have any doubts please ask on slack or github discussion.

Comment (ramakotareddy) [reply compiler/codegen/BinaryOpInsn.cpp:102]: @Ruvi, As per Ballerina Spec and discussion with ayonam, floats are ordered instructions. 
Below cases are printing NaN, Please check once.

    float nanVal = 0.0 / 0.0;
    printf64(nanVal);
    float a = 0.0 % 0.0;
    printf64(a);

Please share small example of any comparison NaN case.

File: compiler/codegen/BinaryOpInsn.cpp (hunk 1 above)
Comment (ramakotareddy) [reply compiler/codegen/BinaryOpInsn.cpp]: Fixed

File: test/float/add_float_negative_values.bal (hunk 3 above)
Comment (ramakotareddy) [reply test/float/add_float_negative_values.bal]: Fixed

File: test/float/greater_equal_than_float.bal (hunk 4 above)
Comment (ramakotareddy) [reply test/float/greater_equal_than_float.bal:16]: Added equal case test

Comment (ruvi-d) [reply test/float/greater_equal_than_float.bal:16]: I meant something like this:
```
    float a = 5.5;
    float b = 5.5;
    print_boolean(a >= b);  // print true
```

File: compiler/codegen/BinaryOpInsn.cpp (hunk 2 above)
Comment (ruvi-d) [reply compiler/codegen/BinaryOpInsn.cpp:102]: I don't know where NaN comparison is required. 
But definitely make the NaN print a test case if it is the same behavior on jBallerina

**Review by ruvi-d — CHANGES_REQUESTED** [review]

File: test/float/greater_equal_than_float.bal (hunk 4 above)
Comment (ramakotareddy) [reply test/float/greater_equal_than_float.bal:16]: @Ruvi, with above test case, always printing false. created new issue #276

File: compiler/codegen/BinaryOpInsn.cpp (hunk 2 above)
Comment (ramakotareddy) [reply compiler/codegen/BinaryOpInsn.cpp:102]: Created New issue as per discussion in today's call and assign to Manu. #279

**Review by ruvi-d — APPROVED** [review]

**Review by KavinduZoysa — APPROVED** [review]

**Pull request merged**